### ISSUE TITLE
Math.random() function not found

### ISSUE THREAD

**Issue by jldevelop** [gpujs/gpu.js#194]
Hi!
I am using few Math functions in my project (Math.floor(), Math.Ceil()...), but it seems that there's not "Math.random()" function found.
When I try to run Math.random() function i get error:
"An error occurred compiling the shaders: ERROR: 0:176: 'random' : no matching overloaded function found
"

**Comment by RationalCoding** [comment]
It hasn't been added yet.

Aside from the deprecated [noise](https://msdn.microsoft.com/en-us/library/bb509629%28v=vs.85%29.aspx) function, there doesn't seem to be any sources of randomness in fragment shaders. :/

We could use arithmetic methods since Math.random() isn't considered secure anyhow.

**Comment by robertleeplummerjr** [comment]
From gitter https://gitter.im/gpujs/gpu.js?at=59cef5e251915d962af9fbf9:

>Robert Plummer @robertleeplummerjr 21:39
yea, random is like the white whale
But we should have it before v1.
Even when implemented, it is still pretty “loose” of a random.
Testing here: https://www.shadertoy.com/view/XtfcRM

**Comment by robertleeplummerjr** [comment]
I'm open to other means as well ^

**Comment by robertleeplummerjr** [comment]
Implementing this now.  Reminds me of Moby Dick.... the "white whale"

**Comment by robertleeplummerjr** [comment]
As taken from https://www.shadertoy.com/view/XtfcRM which is taken from https://stackoverflow.com/a/17479300 this is what I currently have:

```glsl
// A single iteration of Bob Jenkins' One-At-A-Time hashing algorithm.
uint hash(uint x) {
  x += ( x << 10u );
  x ^= ( x >>  6u );
  x += ( x <<  3u );
  x ^= ( x >> 11u );
  x += ( x << 15u );
  return x;
}

// Construct a float with half-open range [0:1] using low 23 bits.
// All zeroes yields 0.0, all ones yields the next smallest representable value below 1.0.
float floatConstruct(uint m) {
  const uint ieeeMantissa = 0x007FFFFFu; // binary32 mantissa bitmask
  const uint ieeeOne = 0x3F800000u;      // 1.0 in IEEE binary32

  m &= ieeeMantissa;                     // Keep only mantissa bits (fractional part)
  m |= ieeeOne;                          // Add fractional part to 1.0

  float f = uintBitsToFloat( m );        // Range [1:2]
  return f - 1.0;                        // Range [0:1]
}

// Pseudo-random value in half-open range [0:1].
float random(float x) {
  return floatConstruct(hash(floatBitsToUint(x)));
}
```

What I'm hoping to prevent is https://medium.com/@betable/tifu-by-using-math-random-f1c308c4fd9d (Note: fantastic article, but I'd hate to be the target of such a well written piece, when I am just trying to make a good tool for others and my greedy self)

The gpu kernel will then have fed in the current time, which is used as the seed to build random on.  Thoughts?

**Comment by dannycalleri** [comment]
First of all, which options do we have? 
In my opinion we should take a scientific approach to it. We could test the statistical properties of the algorithms
we want to consider and take a decision once we have enough data.
While it's true that JavaScript's Math.random() sucks, even Java's one isn't bullet-proof and it's a Linear Congruent Generators, 
one of the simplest forms of PRNG, so it depends on the application.
In our case the problem is that we need an algorithm that is as stateless as possible:
* https://math.stackexchange.com/questions/337782/pseudo-random-number-generation-on-the-gpu
* https://developer.nvidia.com/gpugems/GPUGems3/gpugems3_ch37.html

**Comment by dannycalleri** [comment]
Hey @robertleeplummerjr , testing some ideas taken from here https://developer.nvidia.com/gpugems/GPUGems3/gpugems3_ch37.html

1. https://www.shadertoy.com/view/Ml2yWt (LCG)
2. https://www.shadertoy.com/view/4lfBzn (LCG + Taus step, as discussed here)

I'm pretty sure that I'm missing something especially on the second one, since the result is not what I expected and contains some artifacts. The next step would be to calculate the statistic properties of those and see which one (including the previous shader toy above) performs better, but first I need some help to figure out how to fix the number 2.
If you ask me why sometimes I do some additional XORs it's because it seems a good way to "combine" some random numbers while keeping the random distribution (an easy way to think about it is XOR's truth table. Given two binary inputs A and B, 50% chance to get 1, 50% chance to get 0).

In my opinion it's also worth discussing to calculate the random numbers using JS native lib at compile time on CPU.

**Comment by robertleeplummerjr** [comment]
The artifacts and then them being covered up have me concerned, but energy is good.  Though not perfect: https://www.shadertoy.com/view/XtfcRM doesn't seem to have any artifacts (I'd like to see a histogram though).  To 
To keep the terminology clear, obtaining values from javascript, there are two: "compile" time, and "run" time.

* Compile time: would potentially "freeze" the random values with the compilation, so once you get 0.345879, you have it... 
![](https://media1.giphy.com/media/3orif7gGXMF9EAhwxa/giphy.gif)
* Run time: the value would be fed in as the kernel runs, so if you needed a thousand random numbers, they would be different for every run.  More dynamic...
![](https://media0.giphy.com/media/iOz3p2txHIo4U/giphy.gif)

Off the cuff, what if we did:
Existing: accurate through js
Proposal, started coding: less-accurate, but usable using something like the above shader-toy items

All suggestions are welcome, and ty for time considering, random can be boring.

**Comment by fuzzie360** [comment]
@robertleeplummerjr 

For the compile time solution, you can avoid freezing the random values.

1) Add parameter to control how many random variables are available to each thread. Perhaps `setEntrophyPoolSize` or something (in GLSL, all loops are fixed length anyway, so you can always calculate the maximum numbers required by counting the number of `Math.random()` calls)
2) Before running the kernel, generate `that many random * output width * output height * output depth` and pass it as the `entrophyArray`
3) On compile time, define `Math.random()` as an array lookup on `entrophyArray`, using the `flattened thread ID * entrophyPoolSize + seed` and increment the global variable `seed` locally every time `Math.random()` is called.

This is of course, if you want cryptographic quality randomness, otherwise the PRNG idea from your shadertoy would be really great! We could have an option for either/or.

**Comment by fuzzie360** [comment]
Oh sorry I missed the "run time" bullet point. Yes, exactly this

**Comment by dannycalleri** [comment]
Hi @fuzzie360 , perfectly understood your idea and seems fine to me 👍 
To me it's not a matter of accuracy (since we know - given the above blog post - that Math.random isn't cryptographically accutarate at all), but I think that the compile time solution will get better with the browsers' JS engines getting better, so it would be good solution that doesn't necessarily mean reinventing the wheel.
What do you think @robertleeplummerjr ? 
If we all agree on that, I'll start coding right away, with some help from you of course 😄

**Comment by robertleeplummerjr** [comment]
Compile time, or run time?

**Comment by dannycalleri** [comment]
Sorry, talking about the compile time one :)

**Comment by robertleeplummerjr** [comment]
If you want to do a proof of concept, by all means!

**Comment by dannycalleri** [comment]
Yep, I just wanted to know what do you think about the approach in general, since you are waaaaaay more knowledgeable than me on the project of course :) And also if you think that my argumentation makes sense to you.
As a next step, can you please give me a few advices to start getting hands dirty?
Given what @fuzzie360 proposed, it would be nice to understand where to start that flow
Thank you.

**Comment by robertleeplummerjr** [comment]
My initial concern is this: If the person developing with gpu.js has gotten to the point they are wanted to start messing with `setEntrophyPoolSize`, then just let them give an input that has all the random values they want.  90% of the time (probably more) the random-ness won't matter, and they can just use `Math.random()` which will use the shadertoy script I have above.  This gives us the flaxibility to add a script (which locally I have 90% done) and the gives the end user the flexibility to define their random values a way they see fit.

**Comment by dannycalleri** [comment]
Ok, at this point I don't understand in which way you'd like to address the problem, or if some help is needed 😄

**Comment by robertleeplummerjr** [comment]
I think it is mostly just a sanity check.  I think vote up or down with reason will do nicely.

**Comment by robertleeplummerjr** [comment]
Also: please don't take my arguments as argumentative 😛, I'm trying for simplicity and the abruptness of that target often can be accidentally rude, sorry for that.

**Comment by robertleeplummerjr** [comment]
I was finally able to get my ideas to code here: https://github.com/gpujs/gpu.js/tree/194-adding-random-support

Though I was not aware that uint isn't supported by webgl 1.  So it looks like random is supported currently only by sending it in via js, until v2 is released, which will have webgl 2.0 support.

**Comment by robertleeplummerjr** [comment]
Was able to get this working in v2, rc1.  And while I'd recommend using the CPU `Math.random()`, especially where randomness is super important, the new solution should suffice for most scenarios: https://github.com/gpujs/gpu.js/releases/tag/2.0.0-rc.1

Closing as already released, although it is pre-release.
